b2sdk/sync/sync.py

```diff
@@ -180,52 +98,98 @@ def count_files(local_folder, reporter):
-    :type no_progress: bool
-    :param max_workers: max number of workers
-    :type max_workers: int
-    :param policies_manager: policies manager object
-    :param dry_run:
-    :type dry_run: bool
-    :param allow_empty_source: if True, do not check whether source folder is empty
-    :type allow_empty_source: bool
-    """
-
-    # For downloads, make sure that the target directory is there.
-    if dest_folder.folder_type() == 'local' and not dry_run:
-        dest_folder.ensure_present()
-
-    if source_folder.folder_type() == 'local' and not allow_empty_source:
-        source_folder.ensure_non_empty()
-
-    # Make a reporter to report progress.
-    with SyncReport(stdout, no_progress) as reporter:
+class KeepOrDeleteMode(Enum):
+    DELETE = 301
+    KEEP_BEFORE_DELETE = 302
+    NO_DELETE = 303
+
+
+class Synchronizer(object):
+    def __init__(
+        self,
+        max_workers,
+        policies_manager=DEFAULT_SCAN_MANAGER,
+        dry_run=False,
+        allow_empty_source=False,
+        newer_file_mode=NewerFileSyncMode.RAISE_ERROR,
+        keep_days_or_delete=KeepOrDeleteMode.NO_DELETE,
+        compare_version_mode=CompareVersionMode.MODTIME,
+        compare_threshold=None,
+        keep_days=None,
+    ):
+        """
+        Initialize synchronizer class and validate arguments
+
+        :param int max_workers: max number of workers
+        :param policies_manager: policies manager object
+        :param bool dry_run: test mode, does not actually transfer/delete when enabled
+        :param bool allow_empty_source: if True, do not check whether source folder is empty
+        :param b2sdk.v1.NewerFileSyncMode newer_file_mode: setting which determines handling for destination files newer than on the source
+        :param b2sdk.v1.KeepOrDeleteMode keep_days_or_delete: setting which determines if we should delete or not delete or keep for `keep_days`
+        :param b2sdk.v1.CompareVersionMode compare_version_mode: how to compare the source and destination files to find new ones
+        :param int compare_threshold: should be greater than 0, default is 0
+        :param int keep_days: if keep_days_or_delete is `b2sdk.v1.KeepOrDeleteMode.KEEP_BEFORE_DELETE`, then this should be greater than 0
+        """
+        self.newer_file_mode = newer_file_mode
+        self.keep_days_or_delete = keep_days_or_delete
+        self.keep_days = keep_days
+        self.compare_version_mode = compare_version_mode
+        self.compare_threshold = compare_threshold or 0
+        self.dry_run = dry_run
+        self.allow_empty_source = allow_empty_source
+        self.policies_manager = policies_manager
+        self.max_workers = max_workers
+        self._validate()
+
+    def _validate(self):
+        if self.compare_threshold < 0:
+            raise InvalidArgument('compare_threshold', 'must be a positive integer')
+
+        if not isinstance(self.newer_file_mode, NewerFileSyncMode):
```

Comment:
use `self.newer_file_mode in NewerFileSyncMode.__members__` instead of isinstance. `isinstance` is really bad.